### PR TITLE
osd/PGLog: delay populating PG log hashmaps

### PR DESCRIPTION
**Pull request by branch-predictor**
When booting up OSD, it loads all PGs and their respective logs. To speed
up processing later, these logs are accompanied by separate unordered_maps
which are also populated during PG load.
Delay that until we actually need to access it, so we don't occupy too much
memory right from start - and when we need it, populate just the map that
we want to use, not all 3 of them at once.
Additionally, removed adding all PGs to interim set, because it is actually
unnecessary and code executed during that loop could be moved to other loop.

On my 4-node dev cluster with 98 osds and 52667 objects, combined RSS memory usage for all nodes during cluster bootup looks like this:

mem_boot

Also, there's a tangible difference when rebalancing:

mem_rebalance

(at the beginning, OSDs on one of nodes are being SIGTERMed one by one, after a while recovery kicks in, when it stabilizes, I turn OSDs back online).

Signed-off-by: Piotr Dałek <piotr.dalek@ts.fujitsu.com>

### CODE REVIEW

**Comment by markhpc** [comment]
Any downside to the delayed map population?  I imagine there's a small latency hit when you need it?

**Comment by markhpc** [comment]
Btw, wanted to add that this is fantastic.  Exactly the kind of thing we need to be on the lookout for!

**Comment by branch-predictor** [comment]
If there is a delay, then it's quite low (sub-milisecond), and just once per PG.

**Comment by branch-predictor** [comment]
I updated this PR with changes to PGLog and IndexLog so sizeof(PGLog) == 608 instead of 624. Not much but nice to have.

**Comment by branch-predictor** [comment]
Oh well, looks like the bot has some issues.

**Comment by dachary** [comment]
Yes :-( I'll fix it right away.

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -2827,31 +2827,25 @@ void OSD::load_pgs()
     derr << "failed to list pgs: " << cpp_strerror(-r) << dendl;
   }
 
-  set<spg_t> pgs;
```
Comment (liewegas) [inline src/osd/OSD.cc]: Can you move the removal of this into a separate patch?  It's logically unrelated to the other change.

File: src/gmock [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit 49beb3bdf05a728afb48dbfbeb1a693ce4c38027
+Subproject commit d701f9d547a03677d612d7af3f212871a427f058
```
Comment (liewegas) [inline src/gmock]: It looks like this patch got combined with some other stuff from master somehow.. and there's also a master merge commit in there.  Rebase and clean that out?

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -2861,13 +2855,13 @@ void OSD::load_pgs()
     dout(10) << "pgid " << pgid << " coll " << coll_t(pgid) << dendl;
     bufferlist bl;
     epoch_t map_epoch = 0;
-    int r = PG::peek_map_epoch(store, pgid, &map_epoch, &bl);
+    ghobject_t pgmeta_oid(pgid.make_pgmeta_oid());
+    int r = PG::peek_map_epoch(store, *it, &map_epoch, &bl, pgmeta_oid);
```
Comment (liewegas) [inline src/osd/OSD.cc]: s/*it/pgid/ for clarity

Comment (liewegas) [inline src/osd/OSD.cc]: ...and calculate pgmeta_oid inside peek_map_epoch

File: src/osd/PGLog.h [hunk 4]
```diff
@@ -161,9 +187,12 @@ struct PGLog {
 
     /// get a (bounded) list of recent reqids for the given object
     void get_object_reqids(const hobject_t& oid, unsigned max,
-			   vector<pair<osd_reqid_t, version_t> > *pls) const {
+			   vector<pair<osd_reqid_t, version_t> > *pls) {
       // make sure object is present at least once before we do an
       // O(n) search.
+      if (!(indexed_data && PGLOG_INDEXED_OBJECTS)) {
```
Comment (liewegas) [inline src/osd/PGLog.h]: & not &&

File: src/osd/ReplicatedPG.cc [hunk 5]
```diff
@@ -1651,7 +1651,7 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     // purposes here it doesn't matter which one we get.
     eversion_t replay_version;
     version_t user_version;
-    bool got = pg_log.get_log().get_request(
+    bool got = const_cast<PGLog::IndexedLog&>(pg_log.get_log()).get_request(
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: I don't think we want const_cast.

Instead, we could make the indexes in PGLog mutable and leave all those methods const. Or, just make get_log() return a non-const ref...

**Comment by liewegas** [comment]
Overall this looks pretty promising!  Please rebase and clean out the weird merge cruft/weirdness that is mixed in... and separate out the patch removing the first load_pgs loop.  Thanks!

**Comment by branch-predictor** [comment]
Thanks for your comments. I will probably push out two new PRs out of this, because manual resolving conflicts stemming from changes merged to master after I posted this PR made it like that (it's the best I could do without breaking things/making it even less readable), and I was afraid you wouldn't like it, honestly. I'll close this PR afterwards.

File: src/osd/OSD.cc (hunk 3 above)
Comment (branch-predictor) [inline src/osd/OSD.cc]: We're reusing it later (passing to pg->read_state()). Or I'm not aware of some side effect going on that changes it before reaching pg->read_state()?

File: src/gmock (hunk 2 above)
Comment (branch-predictor) [inline src/gmock]: I'll probably just make two (or even three) separate PRs out of this. Should be easier to merge and debug/blame later.

File: src/osd/OSD.cc (hunk 1 above)
Comment (branch-predictor) [inline src/osd/OSD.cc]: Yes, I'll do it.

File: src/osd/PGLog.h (hunk 4 above)
Comment (branch-predictor) [inline src/osd/PGLog.h]: Whoops! Sorry...

File: src/osd/OSD.cc (hunk 1 above)
Comment (branch-predictor) [inline src/osd/OSD.cc]: Moved to https://github.com/ceph/ceph/pull/6277

**Comment by branch-predictor** [comment]
Part of this patch also moved to https://github.com/ceph/ceph/pull/6279

**Comment by branch-predictor** [comment]
Moved remaining changes to #6425.
